scripts/backup-restore.sh

```diff
@@ -0,0 +1,401 @@
+  echo "Targets: opensearch | harbor | rclone | velero"
+  exit 1
+}
+
+# Sets a timestamp log before every message, also colors it purple to make it easier to distinct from rest of the text because why not
+PURPLE='\033[0;35m'
+NOCOLOR='\033[0m'
+log() {
+  echo -e "[${PURPLE}$(date +'%Y-%m-%d %H:%M:%S')${NOCOLOR}]: $1"
+}
+
+# allows the script to be run from anywhere
+SCRIPT_DIR=$(cd -- "$(dirname -- "${BASH_SOURCE[0]}")" &>/dev/null && pwd)
+REPO_ROOT=$(realpath "$SCRIPT_DIR/..")
+CK8S_CMD="$REPO_ROOT/bin/ck8s"
+
+if [ "$#" -ne 2 ]; then
+  usage
+fi
+
+ACTION="$1"
+TARGET="$2"
+
+run_backup() {
+  case "$1" in
+  opensearch)
+    log "Starting OpenSearch backup..."
+
+    local user="admin"
+    local password
+    password=$(sops -d "${CK8S_CONFIG_PATH}/secrets.yaml" | yq '.opensearch.adminPassword')
+    local os_url
+    os_url="https://opensearch.$(yq '.global.opsDomain' "${CK8S_CONFIG_PATH}/common-config.yaml")"
+    log "OpenSearch URL: ${os_url}"
+
+    log "Discovering snapshot repository name..."
+    local snapshot_repo
+    snapshot_repo=$(curl -kL -s -u "${user}:${password}" "${os_url}/_cat/repositories?v" | awk 'NR==2 {print $1}')
+    if [ -z "$snapshot_repo" ]; then
+      log "ERROR: Could not find OpenSearch snapshot repository."
+      exit 1
+    fi
+    log "Found snapshot repository: ${snapshot_repo}"
+
+    local snapshot_name
+    snapshot_name="manual-snapshot-$(date +%Y%m%d%H%M%S)"
+    log "Taking snapshot: ${snapshot_name}"
+    curl -kL -u "${user}:${password}" -X PUT "${os_url}/_snapshot/${snapshot_repo}/${snapshot_name}" -H 'Content-Type: application/json' -d'
+      {
+        "indices": "*,-.opendistro_security",
+        "include_global_state": false
+      }
+      '
+
+    log "Waiting for snapshot to complete..."
+    local status=""
+    local retries=60 # 30 min timeout
+    while [ $retries -gt 0 ]; do
+      status=$(curl -kL -s -u "${user}:${password}" "${os_url}/_snapshot/${snapshot_repo}/${snapshot_name}" | jq -r '.snapshots[0].state')
+      log "Current snapshot status: ${status}"
+
+      if [ "$status" = "SUCCESS" ]; then
+        log "OpenSearch backup successfully completed."
+        return
+      elif [ "$status" = "FAILED" ] || [ "$status" = "PARTIAL" ]; then
+        log "ERROR: OpenSearch snapshot failed with status: ${status}"
+        exit 1
+      fi
+
+      sleep 30
+      retries=$((retries - 1))
+    done
+
+    log "ERROR: Timeout waiting for OpenSearch snapshot to complete."
+    exit 1
+    ;;
+  harbor)
+    log "Starting Harbor backup..."
+    local backup_job_name
+    backup_job_name="harbor-backup-manual-$(date +%Y%m%d%H%M%S)"
+    log "Creating on-demand backup job: $backup_job_name"
+
+    "$CK8S_CMD" ops kubectl sc -n harbor create job "$backup_job_name" --from=cronjob/harbor-backup-cronjob
+
+    log "Waiting for backup job to complete..."
+    "$CK8S_CMD" ops kubectl sc -n harbor wait --for=condition=complete "job/$backup_job_name" --timeout=30m
```

Comment:
Question: Will this not return error if timed out or if job failed? If so we should check status

```suggestion
    "$CK8S_CMD" ops kubectl sc -n harbor wait --for=condition=complete "job/$backup_job_name" --timeout=30m
    
    if [ $? -ne 0 ]; then
      log "ERROR: Harbor backup job failed"
      exit 1
    fi
```

Reply:
There's a set -e command at the top of the script that would exit the script if it failed. But you are 100% right I forgot to add a status check so it would just exit immediately without saying anything if failed.

Since there's a set -e, I believe your suggestion would fail as well since it would exit immediately and not do the if command? I added a `if !` check instead for the kubectl wait command, that should error out with a message if it fails I believe.